lectures/zksnark.tex

```diff
@@ -0,0 +1,408 @@
+    defined for $u, v \in \mathbb{V}$ as a linear combination:
+    \[ \< u, v \> = \sum_{i=1}^{n} u_i v_i \]
+    With next properties:
+    \begin{itemize}
+        \item Symmetry: $\forall u, v \in \mathbb{V}: \< u, v \> = \overline{\< v, u \>}$
+        \item Linearity in the first argument: $\forall u, v, w \in \mathbb{V}, \forall a \in 
+        \mathbb{F}: \< au + v, w \> = a \< u, w \> + \< v, w \>$
+        \item Positivity: $\forall u \in \mathbb{V}: \< u, u \> \geq 0$ and 
+        $\< u, u \> = 0 \Leftrightarrow u = 0$
+    \end{itemize}
+\end{definition}
+
+
+\begin{example}
+    Let $u, v$ are vectors over the real number $\mathbb{R}$, where
+    \[u = (1, 2, 3), v = (2, 4, 3)\], then:
+    $\<u, v\> = \sum_{i=1}^{3}u_iv_i = 2*1 + 2*4 + 3*3 = 2 + 8 + 9 = 19$
+\end{example}
+
+With knowledge of the inner product of two vectors, we can now formulate a definition of the 
+constraint in the context of an R1CS.
+
+\begin{definition}
+    Each \textbf{constraint} in the Rank-1 Constraint System must be in the form:
+    \[ \<a, w\> \times \<b, w\> = \<c, w\> \]
+    Where $\mathbf{w}$ is a vector containing all the \textit{input}, \textit{output}, and 
+    \textit{intermediate} variables involved in the computation. The vectors $\mathbf{a}$, 
+    $\mathbf{b}$, and $\mathbf{c}$ are vectors of coefficients corresponding to these variables, 
+    and they define the relationship between the linear combinations of $\mathbf{w}$ on the 
+    left-hand side and the right-hand side of the equation.
+\end{definition}
+
+\begin{example}
+    Let's get back to our example from the previous chapter about the arithmetic circuits.
+    \[ r = x_1 \times x_2 \]
+    The constraint is:
+    \[(a_1w_1 + a_2w_2 + a_3w_3)(b_1w_1 + b_2w_2 + b_3w_3) = c_1w_1 + c_2w_2 + c_3w_3\]
+    Coefficients and witness vectors are:
+    $w = (r, x_1, x_2), a = (0, 1, 0), b = (0, 0, 1), c = (1, 0, 0)$. Therefore:
+    \[ (0w_1 + 1w_2 + 0w_3)(0w_1 + 0w_2 + 1w_3) = (1w_1 + 0w_2 + 0w_3) \]
+    \[ w_2 \times w_3 = w_1 \]
+    \[ x_1 \times x_2 = r \]
+\end{example}
+
+
+\begin{example}
+    And a more complex example. Remember that we want to verify each computational step.
+
+    \begin{lstlisting}[language=C,numbers=none]
+    if (x1) {
+        return x2 * x3
+    } else {
+        return x2 + x3
+    }
+    \end{lstlisting}
+
+    We know that it can be expressed as:
+    \[ r = x_1 \times (x_2 \times x_3) + (1 - x_1) \times (x_2 + x_3) \]
+
+    However, one important consideration was overlooked. If $x_1$ is neither $0$ nor $1$, it implies
+    that something else is being computed instead of the desired program. Since we need to add a
+    restriction for $x_1$: $x_1 \times (1 - x_1) = 0$, this effectively checks that $x_1$ is binary.
+
+    The next constraints can be build:
+    \begin{align*}
+        x_1 \times x_1 &= x_1 \quad \text{(binary check)} \tag{1} \\
+        x_2 \times x_3 &= mult \tag{2} \\
+        x_1 \times mult &= selectMult \tag{3} \\
```

Comment:
Either do `\mathsf{mult}` or use `t_1` or `\mathsf{temp}_1` to denote intermediate temporary variables